jax/experimental/export/_shape_poly_decision.py

```diff
@@ -0,0 +1,412 @@
+        #   if m_c > 0. Similar rules for when m_c < 0 and for MIN.
+        op1, op2 = m_a.operands
+        rest1 = _DimExpr._merge_sorted_terms(e, i + 1, 1,
+                                             op1._monomials_sorted, 0, m_c)
+        rest2 = _DimExpr._merge_sorted_terms(e, i + 1, 1,
+                                             op2._monomials_sorted, 0, m_c)
+        rest1_lb, rest1_ub = self._bounds_for_sorted_terms(scope, rest1, 0, None)
+        rest2_lb, rest2_ub = self._bounds_for_sorted_terms(scope, rest2, 0, None)
+        like_max = (m_c > 0 if m_a.operation == _DimAtom.MAX else m_c < 0)
+        if like_max:
+          lb = max(lb, max(rest1_lb, rest2_lb))
+          ub = min(ub, max(rest1_ub, rest2_ub))
+        else:
+          lb = max(lb, min(rest1_lb, rest2_lb))
+          ub = min(ub, min(rest1_ub, rest2_ub))
+        if stop_early is not None and stop_early(lb, ub): return (lb, ub)
+
+    return lb, ub
+
+def _add_internal_constraints(decision: _DecisionByElimination, m: _DimMon, scope: SymbolicScope):
+  """Adds the internal constraints for the monomial `m`."""
+  if m in decision._processed_for_internal_constraints: return
+  decision._processed_for_internal_constraints.add(m)
+  m_e = _DimExpr.from_monomial(m, 1, scope)  # m as a _DimExpr
+  a = m.to_atom()
+  if a is None:
+    # This is a multiplication of atoms. Try to compute bounds based on
+    # the bounds of the atoms.
+    bounds = []
+    for a, exp in m.items():
+      a_l, a_u = decision.bounds(_DimExpr.from_monomial(_DimMon.from_atom(a, 1),
+                                                        1, scope), None)
+      assert a_l <= a_u
+      bounds.append((a_l ** exp, a_u ** exp))
+
+    candidate_bounds = [math.prod(atom_bounds)
+                        for atom_bounds in itertools.product(*bounds)]
+    m_l = min(*candidate_bounds)
+    m_u = max(*candidate_bounds)
+    decision.add_constraint(m_e, m_l)
+    decision.add_constraint(m_u, m_e)
+    return
+
+  # It is an atom, is it a variable?
+  if (v := a.to_var()) is not None:
+    decision.add_constraint(m_e, 1)  # v >= 1
+    return
+
+  if a.operation == _DimAtom.MOD:
+    op1, op2 = a.operands
+    op2_b_l, op2_b_u = decision.bounds(op2, _stop_early_for_geq0)
+    if op2_b_l > 0:  # positive divisor
+      decision.add_constraint(m_e, 0)  # m >= 0
+      decision.add_constraint(op2 - 1, m_e)  # m <= op2 - 1
+      decision.add_constraint(op2_b_u - 1, m_e)
+    elif op2_b_u < 0:  # negative divisor
+      decision.add_constraint(m_e, op2 + 1)  # m >= op2 + 1
+      decision.add_constraint(m_e, op2_b_l + 1)
+      decision.add_constraint(0, m_e)  # m <= 0
+    return
+
+  if a.operation == _DimAtom.FLOORDIV:
+    op1, op2 = a.operands
+    (op1_l, op1_u) = decision.bounds(op1, None)
+    (op2_l, op2_u) = decision.bounds(op2, None)
+
+    def math_floor_with_inf(a: float, b: float):  # math.floor, but aware of inf
```

Comment:
Maybe pull this to the top level to avoid taking extra space here (the function is already quite long)?

Also, nit: this seems to be `floordiv` rather than `floor`, i.e. `floor(a/b)`.